core/deadline_test.go

```diff
@@ -29,23 +29,18 @@ func TestDeadliner(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
-	const lateFactor = 1
-	startTime := time.Now()
-
-	deadlineFunc := func(duty core.Duty) time.Time {
-		duration := time.Second
-		if duty.Type == core.DutyExit {
-			// Do not timeout exit duties.
-			return time.Date(9999, 1, 1, 0, 0, 0, 0, time.UTC)
+	deadlineFuncProvider := func() func(duty core.Duty) time.Time {
+		return func(duty core.Duty) time.Time {
+			if duty.Type == core.DutyExit {
+				// Do not timeout exit duties.
+				return time.Date(9999, 1, 1, 0, 0, 0, 0, time.UTC)
+			}
+
+			return time.Now().Add(time.Millisecond * time.Duration(duty.Slot))
```

Comment:
```suggestion
	deadlineFuncProvider := func(duty core.Duty) time.Time {
			if duty.Type == core.DutyExit {
				// Do not timeout exit duties.
				return time.Date(9999, 1, 1, 0, 0, 0, 0, time.UTC)
			}

			return time.Now().Add(time.Millisecond * time.Duration(duty.Slot))
```

Reply:
can add custom deadlineFunc which will not depend of any time assumptions and deadline duties in desired order

Reply:
Also add tests to ignore expired duties

Reply:
> can add custom deadlineFunc which will not depend of any time assumptions and deadline duties in desired order

yes, but i don't think it's required in this test as duties are received in the same order as the order in which they are added

Reply:
> Also add tests to assert ignored expired duties

yes, we do need to also add some expired duties to test that flow, i'm adding that